util/update-pcluster-configs.py

```diff
@@ -0,0 +1,376 @@
+import collections
+import json
+import logging
+import re
+from abc import ABC, abstractmethod
+from urllib.request import urlopen
+
+import argparse
+import boto3
+from botocore.exceptions import ClientError, EndpointConnectionError
+
+logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s [%(name)s] %(message)s")
+
+PARTITIONS = ["commercial", "china", "govcloud"]
+CONFIG_FILES = ["instances", "feature_whitelist"]
+PARTITION_TO_MAIN_REGION = {"commercial": "us-east-1", "govcloud": "us-gov-west-1", "china": "cn-north-1"}
+FILE_TO_S3_PATH = {"instances": "instances/instances.json", "feature_whitelist": "features/feature_whitelist.json"}
+
+
+class S3JsonDocumentManager:
+    def __init__(self, region, credentials=None):
+        self._region = region
+        self._credentials = credentials or {}
+
+    def download(self, s3_bucket, document_s3_path):
+        try:
+            s3 = boto3.resource("s3", region_name=self._region, **self._credentials)
+            instances_file_content = s3.Object(s3_bucket, document_s3_path).get()["Body"].read()
+            return json.loads(instances_file_content)
+        except Exception as e:
+            logging.error(
+                "Failed when downloading file %s from bucket %s in region %s with error %s",
+                document_s3_path,
+                s3_bucket,
+                self._region,
+                e,
+            )
+            raise
+
+    def upload(self, s3_bucket, s3_key, document_data, dryrun=True):
+        try:
+            if not dryrun:
+                object = boto3.resource("s3", region_name=self._region, **self._credentials).Object(s3_bucket, s3_key)
+                object.put(Body=json.dumps(document_data), ACL="public-read")
+            else:
+                logging.info(
+                    "Dryrun mode enabled. The following file would have been uploaded to %s/%s:\n%s",
+                    s3_bucket,
+                    s3_key,
+                    json.dumps(document_data),
+                )
+        except Exception as e:
+            logging.error(
+                "Failed when uploading file %s to bucket %s in region %s with error %s",
+                s3_key,
+                s3_bucket,
+                self._region,
+                e,
+            )
+            raise
+
+    @staticmethod
+    def validate_document(old_doc, new_doc):
+        try:
+            import dictdiffer
+
+            logging.info("Found following diffs: %s", list(dictdiffer.diff(old_doc, new_doc)))
+        except ImportError:
+            logging.warning("Install dictdiffer if you want to display diffs: pip install dictdiffer")
+
+        if S3JsonDocumentManager._assert_document_is_included(old_doc, new_doc):
+            raise Exception("The new configuration file does not include the old entries.")
```

Comment:
Could we have a quick chat about this? I am not sure I got the intention

Reply:
There is no intention :D forgot to change it when I made _assert_document_is_included raise on failures. Fixing

Reply:
:+1: